controlplane/rosa/controllers/rosacontrolplane_controller.go

```diff
@@ -352,6 +365,115 @@ func (r *ROSAControlPlaneReconciler) reconcileDelete(ctx context.Context, rosaSc
 	return ctrl.Result{}, nil
 }
 
+func (r *ROSAControlPlaneReconciler) reconcileKubeconfig(ctx context.Context, rosaScope *scope.ROSAControlPlaneScope, rosaClient *rosa.RosaClient, cluster *cmv1.Cluster) (ctrl.Result, error) {
+	rosaScope.Debug("Reconciling ROSA kubeconfig for cluster", "cluster-name", rosaScope.RosaClusterName())
+
+	clusterRef := client.ObjectKeyFromObject(rosaScope.Cluster)
+	kubeconfigSecret, err := secret.GetFromNamespacedName(ctx, r.Client, clusterRef, secret.Kubeconfig)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return ctrl.Result{}, fmt.Errorf("failed to get kubeconfig secret: %w", err)
+		}
+	}
+
+	password, err := r.reconcileClusterAdminPassword(ctx, rosaScope)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to reconcile cluster admin password secret: %w", err)
+	}
+
+	clusterName := rosaScope.RosaClusterName()
+	userName := fmt.Sprintf("%s-capi-admin", clusterName)
+	apiServerUrl := cluster.API().URL()
+
+	err = rosaClient.CreateAdminUserIfNotExist(cluster.ID(), userName, password)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	clientConfig := &restclient.Config{
+		Host:     apiServerUrl,
+		Username: userName,
+	}
+	token, err := rosa.RequestToken(apiServerUrl, userName, password, clientConfig)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to request token: %w", err)
+	}
+
+	contextName := fmt.Sprintf("%s@%s", userName, clusterName)
+	cfg := &api.Config{
+		APIVersion: api.SchemeGroupVersion.Version,
+		Clusters: map[string]*api.Cluster{
+			clusterName: {
+				Server: apiServerUrl,
+			},
+		},
+		Contexts: map[string]*api.Context{
+			contextName: {
+				Cluster:  clusterName,
+				AuthInfo: userName,
+			},
+		},
+		CurrentContext: contextName,
+		AuthInfos: map[string]*api.AuthInfo{
+			userName: {
+				Token: token.AccessToken,
+			},
+		},
+	}
+	out, err := clientcmd.Write(*cfg)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to serialize config to yaml: %w", err)
+	}
+
+	if kubeconfigSecret != nil {
+		// update existing kubeconfig secret
+		kubeconfigSecret.Data[secret.KubeconfigDataName] = out
+		if err := r.Client.Update(ctx, kubeconfigSecret); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to update kubeconfig secret: %w", err)
+		}
+	} else {
+		// create new kubeconfig secret
+		controllerOwnerRef := *metav1.NewControllerRef(rosaScope.ControlPlane, rosacontrolplanev1.GroupVersion.WithKind("ROSAControlPlane"))
+		kubeconfigSecret = kubeconfig.GenerateSecretWithOwner(clusterRef, out, controllerOwnerRef)
+		if err := r.Client.Create(ctx, kubeconfigSecret); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to create kubeconfig secret: %w", err)
+		}
+	}
+
+	rosaScope.ControlPlane.Status.Initialized = true
+	// requeue to refresh access token when it expires.
+	return ctrl.Result{RequeueAfter: token.ExpiresIn}, nil
```

Comment:
What's the usual expiration here? Is there a way that we can watch the cluster for updates instead of a long requeue?

Reply:
Default is 1 day. I am not aware of any way to watch the cluster for that. Do you have any ideas?

Reply:
Usually we set default reconciliation of all resources every 10 minutes, so we might just want to return here isntead of requeueing

Reply:
didn't know about the syncPeriod. that makes things much easier.